src/functional/iterator/transforms/pass_manager.py

```diff
@@ -7,34 +9,45 @@
 
 
 def apply_common_transforms(
-    ir,
+    ir: ir.Node,
     use_tmps=False,
     offset_provider=None,
     register_tmp=None,
     unroll_reduce=False,
+    common_subexpression_elimination=True,
 ):
     ir = InlineFundefs().visit(ir)
     ir = PruneUnreferencedFundefs().visit(ir)
     ir = NormalizeShifts().visit(ir)
-    ir = InlineLambdas().visit(ir)
-    if not use_tmps:
-        ir = InlineLifts().visit(ir)
-    ir = InlineLambdas().visit(ir)
+    for _ in range(10):
+        inlined_ir = InlineLifts.apply(ir, preserve_shift_count=use_tmps)
+        inlined_ir = InlineLambdas.apply(inlined_ir)
+        if inlined_ir == ir:
+            break
+        ir = inlined_ir
+    else:
+        raise RuntimeError("Inlining lift and lambdas did not converge.")
```

Comment:
Maybe we can merge this separately in case it fixes the problem in Sam's stencil. I'll try it.

Reply:
It does fix the problem. I'll make a PR.